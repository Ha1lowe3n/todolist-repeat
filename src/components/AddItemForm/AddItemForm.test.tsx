import { render, fireEvent, screen } from "@testing-library/react";
import { act } from "react-dom/test-utils";

import { AddItemForm } from "./";

// props
const addItem = jest.fn();

describe("AddItemForm component", () => {
    it("AddItemForm should render", () => {
        const { getByRole } = render(<AddItemForm addItem={addItem} />);

        expect(getByRole("textbox")).toBeInTheDocument();
        expect(getByRole("button", { name: "+" })).toBeInTheDocument();
    });

    describe("snapshots", () => {
        it("snapshot component", () => {
            const component = render(<AddItemForm addItem={addItem} />);

            expect(component).toMatchSnapshot();
        });

        it("snapshot component with error", () => {
            const component = render(<AddItemForm addItem={addItem} />);
            const btn = screen.getByRole("button", { name: "+" });

            fireEvent.click(btn);

            expect(component).toMatchSnapshot();
        });
    });

    it("AddItem works", () => {
        const { getByPlaceholderText, getByRole } = render(
            <AddItemForm addItem={addItem} />
        );

        const input = getByPlaceholderText(
            "Введите что-нибудь..."
        ) as HTMLInputElement;
        fireEvent.change(input, { target: { value: "$23.0" } });

        const btn = getByRole("button", { name: "+" });
        fireEvent.click(btn);

        expect(addItem).toHaveBeenCalledTimes(1);
    });

    describe("Input", () => {
        it("input change", () => {
            const { getByPlaceholderText } = render(
                <AddItemForm addItem={addItem} />
            );

            const input = getByPlaceholderText(
                "Введите что-нибудь..."
            ) as HTMLInputElement;
            fireEvent.change(input, { target: { value: "$23.0" } });

            expect(input.value).toBe("$23.0");
        });

        it("clear input", () => {
            const { getByPlaceholderText, getByRole } = render(
                <AddItemForm addItem={addItem} />
            );

            const btn = getByRole("button", { name: "+" });
            const input = getByPlaceholderText(
                "Введите что-нибудь..."
            ) as HTMLInputElement;

            fireEvent.change(input, { target: { value: "$23.0" } });
            expect(input.value).toBe("$23.0");
            fireEvent.click(btn);
            expect(input.value).toBe("");
        });
    });

    describe("Error", () => {
        it("show error", () => {
            const { getByRole, queryByText } = render(
                <AddItemForm addItem={addItem} />
            );

            const btn = getByRole("button", { name: "+" });

            expect(queryByText("Title is required")).toBeNull();

            fireEvent.click(btn);
            expect(queryByText("Title is required")).toBeInTheDocument();
        });

        it("error disappears after a while", async () => {
            const { getByRole, queryByText } = render(
                <AddItemForm addItem={addItem} />
            );

            const btn = getByRole("button", { name: "+" });

            expect(queryByText("Title is required")).toBeNull();

            fireEvent.click(btn);
            expect(queryByText("Title is required")).toBeInTheDocument();

            await act(async () => {
                await new Promise<void>((resolve) => {
                    setTimeout(() => {
                        expect(queryByText("Title is required")).toBeNull();
                        resolve();
                    }, 2500);
                });
            });
        });

        it("input change className when error", () => {
            const { getByRole } = render(<AddItemForm addItem={addItem} />);

            const input = getByRole("textbox") as HTMLInputElement;
            const btn = getByRole("button", { name: "+" });

            fireEvent.click(btn);
            expect(input).toHaveClass("error");
        });
    });
});
