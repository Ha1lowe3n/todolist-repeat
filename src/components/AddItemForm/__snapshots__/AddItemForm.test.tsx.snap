// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddItemForm component snapshots snapshot component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="inputTitleTask"
      >
        <input
          class=""
          placeholder="Введите что-нибудь..."
          value=""
        />
        <button
          data-testid=""
        >
          +
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="inputTitleTask"
    >
      <input
        class=""
        placeholder="Введите что-нибудь..."
        value=""
      />
      <button
        data-testid=""
      >
        +
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`AddItemForm component snapshots snapshot component with error 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="inputTitleTask"
      >
        <input
          class="error"
          placeholder="Введите что-нибудь..."
          value=""
        />
        <button
          data-testid=""
        >
          +
        </button>
        <div
          class="errorMessage"
        >
          Title is required
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="inputTitleTask"
    >
      <input
        class="error"
        placeholder="Введите что-нибудь..."
        value=""
      />
      <button
        data-testid=""
      >
        +
      </button>
      <div
        class="errorMessage"
      >
        Title is required
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
